drv/sidecar-front-io/src/leds.rs

```diff
@@ -0,0 +1,320 @@
+        controller: LedController::Left,
+        output: 11,
+    },
+    // Port 22
+    LedLocation {
+        controller: LedController::Left,
+        output: 14,
+    },
+    // Port 23
+    LedLocation {
+        controller: LedController::Left,
+        output: 12,
+    },
+    // Port 24
+    LedLocation {
+        controller: LedController::Right,
+        output: 3,
+    },
+    // Port 25
+    LedLocation {
+        controller: LedController::Right,
+        output: 1,
+    },
+    // Port 26
+    LedLocation {
+        controller: LedController::Right,
+        output: 7,
+    },
+    // Port 27
+    LedLocation {
+        controller: LedController::Right,
+        output: 5,
+    },
+    // Port 28
+    LedLocation {
+        controller: LedController::Right,
+        output: 9,
+    },
+    // Port 29
+    LedLocation {
+        controller: LedController::Right,
+        output: 11,
+    },
+    // Port 30
+    LedLocation {
+        controller: LedController::Right,
+        output: 14,
+    },
+    // Port 31
+    LedLocation {
+        controller: LedController::Right,
+        output: 12,
+    },
+    // System
+    LedLocation {
+        controller: LedController::Left,
+        output: 23,
+    },
+];
+
+impl Leds {
+    pub fn new(
+        left_controller: &I2cDevice,
+        right_controller: &I2cDevice,
+    ) -> Self {
+        Self {
+            controllers: [
+                Pca9956B::new(left_controller),
+                Pca9956B::new(right_controller),
+            ],
+        }
+    }
+
+    /// Set the current to whatever DEFAULT_LED_CURRENT is
+    pub fn initialize_current(&self) -> Result<(), Error> {
+        self.set_current(DEFAULT_LED_CURRENT)?;
+
+        Ok(())
+    }
+
+    /// Set the current to `value`
+    pub fn set_current(&self, value: u8) -> Result<(), Error> {
+        for (_i, controller) in self.controllers.iter().enumerate() {
+            controller.set_iref_all(value)?;
+        }
+
+        Ok(())
+    }
+
+    /// Turns on the System LED to a PWM value of DEFAULT_LED_PWM
+    pub fn turn_on_system_led(&self) -> Result<(), Error> {
+        self.controllers[LED_MAP[SYSTEM_LED_IDX].controller as usize]
+            .set_a_led_pwm(LED_MAP[SYSTEM_LED_IDX].output, DEFAULT_LED_PWM)?;
+
+        Ok(())
+    }
+
+    /// Takes a `mask` of which ports need their LEDs turned on
+    pub fn update_led_state(&self, mask: u32) -> Result<(), Error> {
+        let mut data_l: [u8; 16] = [0; 16];
+        let mut data_r: [u8; 16] = [0; 16];
+
+        for i in 0..32 {
+            let bit_mask: u32 = 1 << i;
+            let pwm_value = if (mask & bit_mask) != 0 {
+                DEFAULT_LED_PWM
+            } else {
+                0
+            };
+
+            if LED_MAP[i].controller == LedController::Left {
+                data_l[LED_MAP[i].output as usize] = pwm_value;
+            } else {
+                data_r[LED_MAP[i].output as usize] = pwm_value;
+            }
+        }
```

Comment:
Since we initialized the arrays to 0, this can be

```suggestion
        for i in 0..32 {
            let bit_mask: u32 = 1 << i;
            if (mask & bit_mask) != 0 {
                let index = LED_MAP[i].output as usize;
                match LED_MAP[i].controller {
                    LedController::Left => data_l[index] = DEFAULT_LED_PWM,
                    LedController::Right => data_r[index] = DEFAULT_LED_PWM,
                }
            }
        }
```

Reply:
🧠